### ISSUE TITLE
Add a way to indicate links are "to do" items

### ISSUE THREAD

**Issue by atopal** [mdn/kuma#6163]
## Current situation
Currently, when an article needs to reference a page that has not yet been written, we still add the `<a>` tags, linking to the intended article's URL. Sometimes these get written right away, sometimes years pass before someone has the opportunity to write that content.

This works fairly well, and gives contributors or potential contributors quick access to the editor through these links.

## The problem
On the down side, some users are confused by this, thinking there's an error of some kind, so we get bug reports about them from time to time.

For multiple reasons, it would be preferable to have a way to label a selection of content as requiring future work to be done. This would let us track and find these to-do items without confusing users.

## Proposed method to mark to-do items
My thinking is that we could use `<mark>` to label content that needs work, with a style that indicates that it's a "to-do" marker:

```
<mark class="mark-todo"
       title="We need to write this landing page">
   <a href="/en-US/docs/Web/API/Some_new_API">
      Some new API
   </a>
</mark>
```
These would be invisible to everyday users (perhaps), but would be highlighted in some way, with a tooltip taken from the `title`, which would be provided while writing the content, to explain the work needing to be done.

## Finding to-do items
Once this is in place, it would be fairly easy to locate and monitor the state of tasks waiting to be done that are labeled in this way. Simply search for uses of the `<mark>` element, and for each of them, you can get the details from the `title` attribute. With that data, you can put it into a spreadsheet or, preferably, a dashboard presenting the most useful information and statistics, with links to create these pages.

**Comment by atopal** [comment]
We discussed removing links for developer.mozilla.org and only displaying them on the wiki.

**Comment by peterbe** [comment]
If a page doesn't exist yet, you have a `class="new"` but no `title` tooltip. 
Example: https://developer.mozilla.org/en-US/docs/Web/Houdini#The_Houdini_APIs
```html
<a class="new" href="/en-US/docs/Web/API/CSS_Properties_and_Values_API" rel="nofollow">CSS Properties and Values API reference</a>
```

I think it would be a mistake to make them invisible. 

Whatever the code is that injects the `new` to the `class`, could also add a `title` tooltip. 

Also, the rather weird thing is that it's still clickable even though it's busted. But it's not busted if the `href` could be re-written to link to the Wiki.

**Comment by peterbe** [comment]
> We discussed removing links for developer.mozilla.org and only displaying them on the wiki.

There are times it makes sense but someone might have typed:

```html
<p>To read more about this, check out: <a href="/does/not/exist/Yet" class="new">This New Thing</a></p>
```

If anything I think it's more valuable to indicate that the page doesn't exist yet. That's actually helping the user. 
Also, if it was a link to the Wiki it'll potentially inspire people to contribute.

**Comment by peterbe** [comment]
Perhaps we can ask some of the experienced editors what they feel about the idea of doing `a.new { display: none }` on the links to pages that don't yet exist. 

A possible concern I have is that you don't notice that they haven't been created yet unless you hop over to `wiki.`.

Thoughts @chrisdavidmills @a2sheppy @wbamberg ?

**Comment by chrisdavidmills** [comment]
@peterbe wouldn't that just cause the links, and the contained link text, to not appear at all on the page? Which would surely make the pages look more broken and confusing than just having red links sat there?

The problem here is that readers see these red links on our pages and go "argh, broken link, there's a bug, must report the problem!" leading to us getting loads of bug reports about things that aren't really bugs.

The upside is that people who want to contribute to MDN can click on these links and then start to write new pages...but only if they have page creation permissions. In reality I don't think seeing the red links ever causes people to click through and go create those new pages. I've worked on MDN fill time for over 6 years, and never done that. Except when I've deliberately created a landing page with broken links on it, with the express intention of going back and filling in thoe pages straight away.

How about just removing the `<a>` elements from around the link text, in cases where the linked pages don't exist, but only on the user version, not on the wiki?

**Comment by peterbe** [comment]
> How about just removing the `<a>` elements from around the link text, in cases where the linked pages don't exist, but only on the user version, not on the wiki?

We can't have different HTML since this is HTML generated inside Django and sent to the wiki and the react front-end as a blob of HTML. Same HTML. 
But what we CAN do with the "user version" (aka. the React site, aka. the read-only site) is to make those links unclickable. E.g. something like this pseudo code:

```javascript
window.onload = function() {
  for (let link in document.querySelectorAll('a.new')) {
    delete a.href
    a.title = "Page does not exist yet"
  }
}
```

There's still stuff we can do differently/better with CSS for these `a.new` selectors. For example, in the breadcrumb, the last element in the breadcrumb `<ul>` is a `<a>` tag but it's styled so it appears like regular plain text. (if that helps)

**Comment by chrisdavidmills** [comment]
Ah, I getcha.

> But what we CAN do with the "user version" (aka. the React site, aka. the read-only site) is to make those links unclickable.

Yeah, I think that might well help. but perhaps make the links look not much like links, with no underline and a more neutral color?

**Comment by schalkneethling** [comment]
> There's still stuff we can do differently/better with CSS for these `a.new` selectors. For example, in the breadcrumb, the last element in the breadcrumb `<ul>` is a `<a>` tag but it's styled so it appears like regular plain text. (if that helps)

We can indeed make the anchor links with class `new` not look like links. Should a user hover over them, however, they will still present as a link.

**Comment by schalkneethling** [comment]
The upside of the CSS approach is not JS that needs to run on each page load and scan the DOM for links marked with `new`

**Comment by peterbe** [comment]
But if we make them not look like links, they're still links. And that's not just true for a human user, it's true for a bot like Googlebot too. 

As concluded before, we can't easily change the raw HTML that Googlebot would encounter but remember that many bots, including Googlebot, do view pages with JS executed. 

Doing a `document.querySelectorAll()` is pretty fast and it's rare that it actually needs to mutate the DOM.

**Comment by schalkneethling** [comment]
> But if we make them not look like links, they're still links. And that's not just true for a human user, it's true for a bot like Googlebot too.

True, but those links also have `rel=nofollow` :)

Easy enough to add the JS though if that works for @chrisdavidmills

**Comment by chrisdavidmills** [comment]
I don't think I'm convinced that what we're suggesting would really be much better. Could I see a prototype in action on a page somewhere, with the ability to compare it to the old version? i.e. before and after?

**Comment by schalkneethling** [comment]
@chrisdavidmills @peterbe Indeed, it does not make that much of a difference. I ran the following against the Houdini page:

```
function disableLinks(links) {
  links.forEach(link => { 
    link.classList.remove("new");
    link.removeAttribute("href");
    link.setAttribute("title", "The page does not exist yet");
  });
}

disableLinks(document.querySelectorAll(".new"));
```

All that does is change the links to look like normal links(i.e. blue and underlined) on MDN. Just removing them entirely is also not an option as they can appear in a multitude of places.

**Comment by schalkneethling** [comment]
This seems slightly better:

```
function disableLinks(links) {
  links.forEach(link => { 
    link.setAttribute("href", "#");
    link.setAttribute("title", "The page does not exist yet");
  });
}

disableLinks(document.querySelectorAll(".new"));
```

Results in the following:

![Screenshot 2019-12-04 at 15 39 04](https://user-images.githubusercontent.com/10350960/70147144-5c7b6c80-16ac-11ea-9496-0b3012b1a85c.png)

**Comment by peterbe** [comment]
Or, to remove them, go to https://developer.mozilla.org/en-US/docs/Web/Houdini#The_Houdini_APIs and open the web console and type:
```
[...document.querySelectorAll('a.new')].forEach(a => a.parentNode.removeChild(a))
```
it causes a bit of a weird extra whitespace. 

And if you go to https://developer.mozilla.org/en-US/docs/Web/API/CSS_Object_Model#CSS_Typed_Object_Model and paste in the same JS snippet in the web console you get this:
<img width="1074" alt="Screen Shot 2019-12-04 at 1 49 06 PM" src="https://user-images.githubusercontent.com/26739/70171522-e0c5f380-169c-11ea-95bc-fefc0713d5ae.png">

I suppose we could work harder on that JS code so it deletes adjacent icons too.

**Comment by peterbe** [comment]
This *might* be unrealistic because my local mysql has less pages than the production DB but this'd be hard to get right if we remove all `a.new` links:
<img width="508" alt="Screen Shot 2019-12-05 at 2 39 05 PM" src="https://user-images.githubusercontent.com/26739/70267820-34077700-176d-11ea-935d-8d6175a80ba1.png">

**Comment by schalkneethling** [comment]
Yup, completely removing them is not an option I reckon.

**Comment by peterbe** [comment]
So, what we know:

1. Removing any `a.new` is probably too weird to attempt. 
2. We can't, in Python, do much because that HTML that Python makes is shipped identically to the wiki and to the read-only site. 
3. If we did this (*) we a ruining the links that are "slightly busted" such as the `<a href="/en-US/docs/Web/Houdini/">Houdini</a>` link (**) in the ["See Also" section](https://developer.mozilla.org/en-US/docs/Web/API/CSS_Object_Model#See_also) which actually worked when you clicked on it. 
4. Followup from (3). I think the first order of business is to, in Python, NOT put `class="new"` on a hyperlink that *would* resolve to a perfectly fine slug if you just omitted the trailing slash. 
5. Red is pretty good styling already. It's the universal color for "negative" or "no".

**How about this...?**

1. I'll look into the Python that put the `class="new"` on that `/en-US/docs/Web/Houdini/` and see if it can be improved so as to be more tolerant with trailing slashes. 
2. @chrisdavidmills you ponder the tooltip title and whether you want to do more visual tricks to these links. 


(*) 
```javascript
[...document.querySelectorAll('#wikiArticle a.new')].forEach(a => {
  a.removeAttribute('href') // makes it not be clickable and no "pointer" cursor when hovering
  a.title = "Page has not yet been created"
}); 
```

(**) This is NOT a KumaScript thing. It's the exact HTML used.

**Comment by chrisdavidmills** [comment]
> you ponder the tooltip title and whether you want to do more visual tricks to these links.

I think the suggested wording is OK - it means what it says. I don't think we should try any further visual tricks for now; let's see if this improves things first.

**Comment by peterbe** [comment]
@schalkneethling Sorry. I saw this was assigned to you but I had a my finger on the trigger and just took a gamble to fix it. Or, at least to get started on a PR.

**Comment by jmswisher** [comment]
Sorry I'm late to this party. How does the proposed solution come across to a screen reader?

**Comment by schalkneethling** [comment]
> Sorry I'm late to this party. How does the proposed solution come across to a screen reader?

The link has a `title` attribute that states `Page has not yet been created.`. We probably want to duplicate that as an `aria-label` so it overrides whatever text is inside the anchor element. What do you think @jmswisher?
